vignettes/compareToBaseR.Rmd

```diff
@@ -0,0 +1,683 @@
+
+```{r}
+head(sentences)
+noun <- "(a|the) ([^ ]+)"
+has_noun <- sentences %>%
+  str_subset(noun) %>%
+  head(5)
+has_noun
+
+## base
+matches <- regexec(pattern = noun, text = has_noun)
+do.call("rbind", regmatches(x = has_noun, m = matches))
+gregexpr(pattern = noun, text = has_noun)
+
+## stringr
+str_match(string = has_noun, pattern = noun)
+str_match_all(string = has_noun, pattern = noun)
+
+```
+
+The output of grouped matches is the full match and then the components of each group. Again, in base R we need to pair `regexec()` and `regmatches()` whereas in stringr we just need one function: `str_match()`. To get all the grouped matches in base R we would need to organize the output of `gregexpr()` whereas stringr's `str_match_all()` streamlines this process.
+
+
+# Manage lengths
+
+## `str_length`
+
+*The length of a string*
+
+To determine the length of a string, base R uses `nchar()` (not to be confused with `length()` which gives the length of vectors, etc.) while stringr uses `str_length()`.
+
+```{r}
+## base
+nchar(letters)
+
+## stringr
+str_length(letters)
+
+```
+
+There are some subtle differences between base and stringr here. `nchar()` requires a character vector, so it will return an error if used on a factor. `str_length()` can handle a factor input.
+
+
+```{r,eval=F}
+## base
+nchar(factor("abc")) ## Error in nchar(factor("abc")) : 'nchar()' requires a character vector
+```
+
+```{r}
+## stringr
+str_length(factor("abc"))
+```
+
+From the documentation: "Technically this [`str_length()`] returns the number of 'code points', in a string. One code point usually corresponds to one character, but not always."
+
+```{r}
+u1 <- "\u00fc"
+u2 <- stringi::stri_trans_nfd(u1)
+
+u1
+u2
+
+## base
+nchar(u1) 
+nchar(u2)
+
+
+## stringr
+str_length(u1)
+str_length(u2)
+
+```
+
+## `str_pad`
+
+*Pad a string*
+
+If you want to pad a string to a certain width, you can use stringr's `str_pad()`. In base R you could use `sprintf()`, but unlike `str_pad()`, `sprintf()` has many other functionalities. 
+
+```{r}
+## stringr
+rbind(
+  str_pad("hadley", 30, "left"),
+  str_pad("hadley", 30, "right"),
+  str_pad("hadley", 30, "both")
+)
+
+## base
+sprintf("%30s", "hadley")
+sprintf("%-30s", "hadley")
+# "both" is not as straightforward
+```
+
+
+## `str_trunc`
+
+*Truncate a character string*
+
+The stringr package provides an easy way to truncate a character string: `str_trunc()`. In base R you would have to do something a bit more complicated, potentially leveraging `substr()`.
+
+```{r}
+x <- "This string is moderately long"
+
+## stringr
+rbind(
+  str_trunc(x, 20, "right"),
+  str_trunc(x, 20, "left"),
+  str_trunc(x, 20, "center")
+)
+
+```
+
+## `str_trim`
+
+*Trim whitespace from a string*
+
+Similarly, stringr provides `str_trim()` to trim whitespace from a string. This is analogous to base R's `trimws()`.
```

Comment:
Might be worth mentioning which version of R `trimws()` was added. IIRC it was fairly recently.